include/abstract_data_store.h

```diff
@@ -76,19 +78,28 @@ template <typename data_t> class AbstractDataStore
     // num_points) to zero
     virtual void copy_vectors(const location_t from_loc, const location_t to_loc, const location_t num_points) = 0;
 
-    // metric specific operations
-
+    // Some datastores like PQ stores need a preprocessing step before querying.
+    // Optionally, a scratch object can be passed in to avoid memory allocations
+    // Default implementation does nothing.
+    // REFACTOR TODO: Currently, we take an aligned_query as parameter, but this
+    // should change and this function should do the necessary alignment.
+    virtual void preprocess_query(const data_t *aligned_query, AbstractScratch<data_t> *query_scratch = nullptr) const;
```

Comment:
Let's clean this up, and do all alignment consistently in a small but easily accessible set of places?